### PR TITLE
Google Calendar - bug fix & refactor

### LINKED ISSUE #2472: [BUG] Google Calendar - New or Update Event (Instant) source fails to renew subscriptions after some period of time
The **New or Update Event (Instant)** Google Calendar source ([see code](https://github.com/PipedreamHQ/pipedream/blob/master/components/google_calendar/sources/new-or-updated-event-instant/new-or-updated-event-instant.mjs)) fails to correctly renew its webhook subscriptions with Google, and the data tied to that subscription is null, which means we can no longer correctly process new events from Google:

Screen Shot 2022-03-22 at 9 38 24 AM

I haven't been able to reproduce the conditions under which it fails. It must fail to make the subscribe API request, and then get into a bad loop. I am sure we can be more defensive about how we renew subscriptions and correct this. At the very least, we should correctly renew the subscription the next time the source runs on its scheduled timer, which does not appear to be happening.

### PR DESCRIPTION
**Pull request by michelle0927**
Addresses a bug in **New or Update Event (Instant)** source and refactors Google Calendar sources.

Resolves #2472

### CODE REVIEW

**Comment by dylburger** [comment]
Thanks for submitting this PR! When we review PRs, we follow the [Pipedream component guidelines](https://pipedream.com/docs/components/guidelines/). If you're not familiar, here's a quick checklist:

- [ ] [Create components to address specific use cases whenever possible](https://pipedream.com/docs/components/guidelines/#component-scope)
- [ ] [Component `key`s should follow the format `app_name_slug-slugified-component-name`](https://pipedream.com/docs/components/guidelines/#component-key-pattern)
- [ ] Components should follow the [standard directory structure](https://pipedream.com/docs/components/guidelines/#folder-structure)
- [ ] [Prefer Node.js client libraries to REST APIs](https://pipedream.com/docs/components/guidelines/#using-apis-vs-client-libraries)
- [ ] When making API requests, [handle pagination](https://pipedream.com/docs/components/guidelines/#pagination) to ensure all data / events are processed
- [ ] [Use `secret` props](https://pipedream.com/docs/components/guidelines/#capturing-sensitive-data) to capture sensitive data
- [ ] Props and methods should be defined in [app files](https://pipedream.com/docs/components/guidelines/#app-files) whenever possible
- [ ] Document methods with [JS Docs](https://pipedream.com/docs/components/guidelines/#methods)
- [ ] [Use `optional` props](https://pipedream.com/docs/components/guidelines/#optional-vs-required-props) whenever possible, and [set a `default` value](https://pipedream.com/docs/components/guidelines/#default-values) where you can
- [ ] Use [async options](https://pipedream.com/docs/components/guidelines/#async-options) to accept user input wherever possible
- [ ] [Create subscription webhooks sources (vs polling sources) whenever possible](https://pipedream.com/docs/components/guidelines/#webhook-vs-polling-sources)
- [ ] [Always emit a `summary` with events](https://pipedream.com/docs/components/guidelines/#emit-a-summary)
- [ ] [Use a built-in deduping strategy wherever possible](https://pipedream.com/docs/components/guidelines/#deduping)
- [ ] When building webhook sources, [follow these best practices](https://pipedream.com/docs/components/guidelines/#webhook-sources)
- [ ] When building polling sources, [follow these best practices](https://pipedream.com/docs/components/guidelines/#polling-sources)

File: components/google_calendar/google_calendar.app.mjs [hunk 1]
```diff
@@ -1,13 +1,27 @@
...
   propDefinitions: {
     calendarId: {
       label: "Calendar ID",
-      description: "Calendar identifier. To retrieve calendar IDs call the calendarList.list method. If you want to access the primary calendar of the currently logged in user, use the \"primary\" keyword.",
+      description: "Calendar identifier. To access the primary calendar of the currently logged in user, use the \"primary\" keyword.",
       type: "string",
+      async options() {
+        const calListResp = await this.calendarList();
+        const calendars = calListResp?.data?.items ?? [];
+        if (calendars && calendars.length) {
```
Comment (vellames-turing) [inline components/google_calendar/google_calendar.app.mjs:14]: Can we use the code below?
```suggestion
        if (calendars?.length) {
```

File: components/google_calendar/sources/common.mjs [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+        id,
+        start,
+      } = event;
+      return {
+        summary,
+        id,
+        ts: +new Date(start.dateTime),
+      };
+    },
+    async processEvents(event) {
```
Comment (vellames-turing) [inline components/google_calendar/sources/common.mjs:53]: There is some `console.log` here at the same line of other statementss

File: components/google_calendar/sources/common.mjs [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+    },
+  },
+  hooks: {
+    async activate() {
+      await this.processEvents({});
+    },
+  },
+  methods: {
+    getIntervalData(event) {
+      const intervalMs = 1000 * (event.interval_seconds || 300); // fall through to default for manual testing
```
Comment (vellames-turing) [inline components/google_calendar/sources/common.mjs:26]: Just checking, if `event.interval_seconds = 0` so the evaluated value will be `300`. Maybe you want to use `??` operator or the idea is to use `300` if `event.interval_seconds = 0 or null`?

**Review by vellames-turing — CHANGES_REQUESTED** [review]
Great refactor! Just some minors comments

**Review by vellames-turing — DISMISSED** [review]
Thank you!

**Comment by vunguyenhung** [comment]
Hello @michelle0927, in order for me to verify, could you resolve conflicts in this PR?

**Comment by michelle0927** [comment]
@vunguyenhung Is the only component that's failing "New or Updated Event (Instant)"? If you have the prop "New Events Only?" set to `true`, it will emit for new events, but not updated events. If you set the prop to `false`, it will emit for new or updated events.
Screen Shot 2022-05-19 at 12 56 03 PM

**Comment by vunguyenhung** [comment]
Thank you @michelle0927 for your explanation. There was only one failed case which is New or Updated Event (Instant), though, I have disable the "New Event" options and I can receive Update event now. Everything is Good! ready for release!

**Comment by vunguyenhung** [comment]
Thank you @michelle0927 for your explanation. There was only one failed case which is New or Updated Event (Instant), though, I have disable the "New Event" options and I can receive Update event now. Everything is Good! ready for release!

Test Report
[Google_Calendar_2472_2113.pdf](https://github.com/PipedreamHQ/pipedream/files/8740792/Google_Calendar_2472_2113.pdf)

**Comment by vunguyenhung** [comment]
Thank you [@michelle0927](https://github.com/michelle0927) for your explanation. There was only one failed case which is New or Updated Event (Instant), though, I have disable the "New Event" options and I can receive Update event now. 

I have also tested the case where events are still emitted after push notification renewal. More detail on the test report

All test cases are passed! Ready for release!

Test report
[Google_Calendar_2472_2128.pdf](https://github.com/PipedreamHQ/pipedream/files/8740913/Google_Calendar_2472_2128.pdf)

**Review by andrewjschuang — APPROVED** [review]

**Pull request merged**